### ISSUE TITLE
Sonoff ZBMINIR2 Binding issue

### ISSUE THREAD

**Issue by kilian-f-za** [Koenkk/zigbee2mqtt#25392]
### What happened?

I have added the Sonos ZBMINIR2 and updated the firmware (1.0.4). Then I created a binding (onoff) and the message "Route successfully created" appeared. However, the device is not switched. I have tried it with several different devices, but it does not work with any of them.
I usw the SMLIGHT SLZB-06M as Coordinator.
![Screenshot_20241231-093150~(1)](https://github.com/user-attachments/assets/f263782f-d64f-492f-b32f-0e65664871aa)

### What did you expect to happen?

The connected device should switch on and off when the input S2 on sonoff is changed. 
With an Aeotec pico it works with the same procedure.

### How to reproduce it (minimal and precise)

Firmware Update then reset the sonoff and reconnect with zigbee2mqtt

### Zigbee2MQTT version

1.40.2

### Adapter firmware version

7.4.5 [GA]

### Adapter

Ember

### Setup

on proxmox lxc Container installed with proxmox helper script

**Comment by bobo-cher** [comment]
I'm afraid bindings do not work on zbminir2 at all. This is hardly a z2m issue but rather a problem (or specific) of the device. I have myself zbminir2 and never managed bindings to work but I haven't tried anything other than z2m. 

Also see [ZBMINIR2 direct zigbee binding](https://www.reddit.com/r/sonoff/comments/1hjuz4h/zbminir2_direct_zigbee_binding/).

An ultimate test would be to setup zbminir2 and other device with Sonoff gateway and their app; then power off the gateway and test if the binding between two devices works.

**Comment by kilian-f-za** [comment]
Thanks for the answer, then I will switch to another device. However, the Sonoff is very small. Direct bindings are very important to me, if the gateway fails at least the light should still come on. It would be great if direct binding worked on many more devices.

**Comment by bobo-cher** [comment]
Maybe you could create a support ticket at Sonoff. They should know users need the feature.

**Comment by ius** [comment]
I've investigated this issue in detail. It's a limitation of the Sonoff firmware indeed (as mentioned by @bobo-cher).

> I'm afraid bindings do not work on zbminir2 at all.
Doesn't matter in practice for the issue at hand, but it _does_ support bindings. It doesn't however address them when sending on/off commands. Sad thing is that it's a one line fix: call `sl_zigbee_af_send_command_unicast_to_bindings`.

@kilian-f-za I believe Sonoff's firmware is somewhat robust in this regard for most use cases as it reverts to switching the relay (even when configured in detached relay mode) when the coordinator becomes unavailable.

I would in fact prefer to ignore coordinator status and rely on bindings for robustness instead.

> Maybe you could create a support ticket at Sonoff. They should know users need the feature.

They told me they had no plans to add this feature. To work around this issue I've started working on my own firmware - I got bindings to work just fine, but all additional fluff (ota, 'factory' reset on 5x toggle) is not yet implemented.

I'm a bit short on time lately, but if I manage to get ota to work I could look into publishing it. Two major drawbacks though: you need to flash it via swd and there's no way to revert to Sonoff firmware..

**Comment by bobo-cher** [comment]
@ius Your project sounds very interesting. I would gladly use one or two of my spare zbminir2 for testing if you are willing to share your work and could provide some instructions for flashing.

My ultimate goal is to use one zbminir2 as sender (connected to wall switch/button) and another zbminir2 as receiver (connected to lamp). It's a shame Sonoff doesn't support that use case out of the box as that would make these small devices such versatile.

That would allow me to transform a simple ceiling light with one wall switch into two ceiling lamps with two independent switches without adding of a new wire. I would use two minis (as senders) in the wall box behind physical switches and two minis (as receivers) on the ceiling.

**Comment by bobo-cher** [comment]
I've created a support ticket with Sonoff and got feedback that direct binding is a known feature request.

I encourage every user interested in the feature to contact Sonoff to vote for the feature. That increases chances we got this. Thanks.

**Comment by ius** [comment]
I've published my ZBMINIR2 firmware to https://github.com/ius/zbminir2 (in progress, experimental!)

Only the essentials required for my binding use case are implemented (now does include OTA and 'reset'). Anyone interested in _ZBMINIR2-with-bindings_ do check it out.

Disclaimer: there _might_ be reliability issues (might just be my n=1 setup) - additional feedback is thus very welcome.

@bobo-cher Also happy to help out on adapting it to your use case if you're willing to give it a try - preferably catch me on Discord (see README).

**Comment by bobo-cher** [comment]
@ius Thank you very much for sharing your project. Appreciate your efforts for publishing of the repository and writing of the docs.

I'll order the required hardware and give it a try. Will get back to you via project issue tracker or Discord. Thanks again!

**Comment by Pferdebockwurst** [comment]
> [@Pferdebockwurst](https://github.com/Pferdebockwurst) where is your source for this?

[Here](https://www.reddit.com/r/sonoff/comments/1hjuz4h/comment/mded4i6/?utm_source=share&utm_medium=mweb3x&utm_name=mweb3xcss&utm_term=1&utm_content=share_button).

**Comment by bjornarfjelldal** [comment]
> > [@Pferdebockwurst](https://github.com/Pferdebockwurst) where is your source for this?
> 
> [Here](https://www.reddit.com/r/sonoff/comments/1hjuz4h/comment/mded4i6/?utm_source=share&utm_medium=mweb3x&utm_name=mweb3xcss&utm_term=1&utm_content=share_button).

Thanks!

**Comment by frankforpresident** [comment]
Does anyone know of an alternative to this switch that supports bindings? I ordered one to test but ran into this limitation. I’ll probably use this module for a regular bulb, but I’d still like to find one that can bind directly to a Zigbee bulb and requires a neutral.

**Comment by andrei-lazarov** [comment]
🎉 **_Update: SONOFF added outgoing binds in v1.0.8 stock firmware._**

---

@frankforpresident last time I checked, the ZBMINIR2 supported incoming binds, but not outgoing binds. (v1.0.4)
Meaning you can control the module with a remote, but you can not use the module as a remote to control something else. 

**You can flash the ZBMINIR2 with the custom firmware mentioned above:** [ius/zbminir2](https://github.com/ius/zbminir2).  
It supports both usecases, but it requires flashing by wire ~~and the project looks stale~~.

**Sadly, there isn't a cheap module that supports binding both ways out-of-the-box**.  
I have asked this before and I tried multiple switches, with no results.

**However, there is a solution:**  
Generic Chinese devices (Tuya) allow installing custom firmware wirelessly!  
So you can get almost any recent cheap module and install [romasku/tuya-zigbee-switch](https://github.com/romasku/tuya-zigbee-switch) with very little effort.  
The project is rapidly growing and it's already feature-rich.  I help new users set up, you're welcome to join the community!

So here are some options that are already ported to romasku's custom firmware and work well:
- [Avatto ZWSM16](https://www.aliexpress.com/item/1005007438984562.html)
- [iHseno modules](https://www.aliexpress.com/item/1005008107698143.html)
- [Aubess modules](https://www.aliexpress.com/item/1005005748264739.html)

**Comment by frankforpresident** [comment]
@andrei-lazarov,

Thank you for this awesome reply. Exactly the information that I needed.

**Comment by ius** [comment]
> It supports both usecases, but it requires flashing by wire and the project looks stale.

The first part is true; unfortunately there's no way around it.

As for the second part, don't feel discouraged from using it because there haven't been any recent commits - basic features (ie. switch + in/output bindings) work and have been rock stable for the past 8 months in my own deployment - thus no additional commits. There's an open issue for supporting pulse switch mode - if I find some time I might add it.

The project you've linked looks nice - while it doesn't have the initial flashing drawback, I'd personally would exercise some caution making sure the hardware is sane from a safety perspective...

**Comment by andrei-lazarov** [comment]
@ius thanks for the project and sorry for calling it stale 🙂 I shout it out whenever I'm given the occasion.  
I'm glad it's stable. I wanted to use it too, but the tiny solder points turned me away.  
The safety concern about Tuya devices is valid though.. any tips on what I should look for?

**Comment by frankforpresident** [comment]
@ius,

You should update the README if the project is stable, because it still says it is experimental which was one of the reasons I decided to skip it

**Comment by NorskNoobing** [comment]
Subscribed to this issue a while ago, and when @andrei-lazarov left his reply here I immediately bought the [Avatto ZWSM16](https://www.aliexpress.com/item/1005007247647375.html). It arrived a couple of days ago, then I created a group in z2m with all my lights, and bound the relay to the lights. I even tested after unplugging my coordinator from power. Safe to say, the zigbee direct bindings definitely work with the custom firmware. Thanks!

Also contacted Sonoff a couple of weeks back for an update, and they said:

> It's a known request, we have forwarded it to the R&D team before. We don't have the estimated time to support this function. We will keep following up on this request, if any updates, we will contact you.

So I wouldn't expect the direct binding support on the ZBMINIR2 any time soon.

**Comment by Hydro8** [comment]
When the coordinator is down, ZBMINIR2 fallback to relay mode even if he’s in detached mode right ?

**Comment by diegosantamarta** [comment]
The only solution right now is to use the firmware @ius said here right?

**Comment by andrei-lazarov** [comment]
> When the coordinator is down, ZBMINIR2 fallback to relay mode even if he’s in detached mode right ?

@Hydro8 Yes. 
I can confirm the SONOFF ZBMINIR2 has some fallback implemented on the original firmware.  
When Z2M is offline, it works in 'relay mode', even if 'detached mode' is enabled. 
But in this case, it reacts to button-presses once every 5-10 seconds. 
(When the network is online, or 'detached mode' is disabled, it reacts to every press)
 
I installed it with a momentary button. I enabled detached mode and stopped Z2M.  
First click toggles the relay. The next 5-10s, it ignores all clicks (like it's attempting to reconnect or something).  
After the timeout you can press it again (once, then wait).

By the way, the detached mode only reports single-clicks as actions (on original fw). No double-clicks, no long-clicks.  
You need an HA automation / blueprint for that. But it seems responsive enough to work.

---

> The only solution right now is to use the firmware @ius said here right?

@diegosantamarta Yes. The information in this thread is still valid. 

My SONOFF is on original fw, version 1.0.4 (20241202), with no further updates provided by Z2M.  
You can bind a remote to its relay. But you can not bind its button to a lightbulb.

If you need that functionality, use custom firmware, as I said in my earlier replies. 
We will eventually support this device on Romasku's firmware too, where you might find more features. 
But the 'initially flash by wire' constraint will remain.

**Comment by diegosantamarta** [comment]
> If you need that functionality, use custom firmware, as I said in my earlier replies.
> We will eventually support this device on Romasku's firmware too, where you might find more features.
> But the 'initially flash by wire' constraint will remain.

Ok, but I understand it's not supported right now?

**Comment by andrei-lazarov** [comment]
It's supported by ius' firmware right now.

**Comment by diegosantamarta** [comment]
Thanks @andrei-lazarov I've bought both ZBMINIR2 and ZWSM16-W1 so I can try on them both one with @ius firmware and the other with @romasku firmware

**Comment by andrazek** [comment]
This is the message I received from Sonoff support:

> The ZBMINIR2 V1.0.8 firmware has begun its phased rollout on our eWeLink platform, adding a binding function that only takes effect after the detach relay mode is enabled. Uploads from third-party platforms will be a little later. You can keep a close eye on the updates.

Can someone using eWeLink upgrade to the latest firmware and try pairing the device with Z2M to confirm whether direct binding works?

**Comment by andrei-lazarov** [comment]
Thanks for sharing the good news @andrazek!
They already pushed the update to Z2M: https://github.com/Koenkk/zigbee-OTA/pull/940
@ius @diegosantamarta @frankforpresident @kilian-f-za 

***SONOFF ZBMINIR2 supports binding both ways on stock firmware v1.0.8!***

I updated a unit and did some tests:
- It sends the **toggle** command to bound lights as soon as you press the button (genOnOff cluster).
  - 'edge' and 'pulse' work as expected
  - 'follow on' and 'follow off' don't work for binds! - same behavior as 'edge'
  - individually toggles each bound device (or each member of the bound group) - unpleasant, but expected behavior (this is how Zigbee toggle works)
- Commands to bound lights are sent only while 'detached relay mode' is active
  - this would conflict with the **fallback mechanism** (disable detached mode when coordinator/Z2M offline), but it seems they **removed it** @Hydro8 
  - I found a workaround to 'toggle relay AND send command' on button press: **bind the sonoff to itself** (it uses endpoint 1 for both input and output)
- It can be bound to individual devices and/or groups
  - bind to device works normally
  - bind to group also toggles itself (sonoff relay) for some reason? 
I double-checked: it's not inside the group and I removed the self-bind
- bind commands are sent even if network is down
- they still did not implement on_with_timed_off as far as I can tell

Of course it's missing more advanced features like hold to dim (genLevelCtrl) and syncing relay state with bound lights.
But this is a great improvement! So far I'm happy with it.

Edit: there are some reports of group binding being unreliable?

**Comment by alexsotoaguilera** [comment]
Just updated my ZBMINIR2 to firmware 1.0.8 and enabled detached relay mode.
Direct bindings from the ZBMINIR2 to individual bulbs work great: the switch sends toggle to each bound light as soon as I press the button and the bulbs stay powered all the time, which is exactly what I wanted.

The only issue I’ve run into is when binding to a group instead of individual lights. With a group binding, the ZBMINIR2 relay starts toggling itself (cutting power to the Zigbee bulbs), even though the device is in detached relay mode and is not part of the group. This doesn’t happen when I bind to the bulbs one by one, only when I bind to the group.

So right now my workaround is:

Use detached relay mode

Bind the ZBMINIR2 directly to each bulb (no group binding)

Keep the bulbs always powered and use “do not disturb” / power‑on settings on the bulbs so they don’t turn on automatically after a power outage.

Apart from the group binding behavior, the new firmware is working very well for my use case.

**Comment by Hydro8** [comment]
Thanks a lot for your feedback on this last firmware. I'm happy that Sonoff put binding option to ZBMINIR2.
This switch seems to be the best one around.

Do you know if it's possible to disable binding from HA ? In case the smart bulb is dead, I would like to easily go back to classic mode for my wife while waiting to change the bulb.

**Comment by andrei-lazarov** [comment]
@Hydro8 yes, you just disable 'detached relay mode'.

**Comment by Hydro8** [comment]
> [@Hydro8](https://github.com/Hydro8) yes, you just disable 'detached relay mode'.

You can do it from an automation or you must edit the setting in z2m ?

**Comment by andrei-lazarov** [comment]
@Hydro8 You can do an automation. It's just a toggle, exposed to HomeAssistant / MQTT.

**Comment by Hydro8** [comment]
> [@Hydro8](https://github.com/Hydro8) You can do an automation. It's just a toggle, exposed to HomeAssistant / MQTT.

Alright thanks ! So I can create a "maintenance mode" where detached relay mode is disabled with one button on the dashboard. Great !

**Comment by uschindler** [comment]
> * bind to group also toggles itself (sonoff relay) for some reason?
>   I double-checked: it's not inside the group and I removed the self-bind

Hi, this is really annoying. It also happens if the ZBMINIR2 is part of a group (I have a group to turn off everything when I leave the house). When you go to detached relay mode and toggle the switch suddenly all other lights in this group also turn on.

This is a bug: When a device is part of a group it should only listen to events for that group and toggle itsself. If a device binds the ONOFF cluster to a group it should toggle that group (and not all groups it is part of).

**Comment by andrei-lazarov** [comment]
@uschindler I couldn't replicate this part. It only toggles the group I bound it to.

**Comment by uschindler** [comment]
> [@uschindler](https://github.com/uschindler) I couldn't replicate this part. It only toggles the group I bound it to.

Hi, sorry. It looks like accidentally I bound the device to the "all lights" group long ago. When removing all group bindings it did no longer happen.

I can confirm your observation in detached relay mode:
- If you bind the device to itsself it behaves as if there's no detached relay mode
- If you bind the device to another device it toggles it

**Comment by uschindler** [comment]
What also works well: If you create a group of all lamps including the ZBMINIR2 and set it to detached relay, you can bind the OnOff cluster to the group. This will toggle the whole group (inclduing itsself) on each button action.

This allows to have the main lamp in the room (non-smart) toggled by the relay and all additional lamps (like indirect lights) toggle together with main lamp when the button is pressed. I implemented that now and works great:

- detached relay mode
- ZBMINIR2 and indirect lights in one group
- ZBMINIR2's OnOff Cluster bound to the group

**Comment by uschindler** [comment]
So actually it is not a fault if the device toggles itsself if it is bound to the group it belongs to. It is basically same behavious as binding to itsself.

**Comment by Pferdebockwurst** [comment]
Can everyone here confirm that the direct binding doesn't support dimming (GenLevelCtrl cluster)?

**Comment by uschindler** [comment]
> Can everyone here confirm that the direct binding doesn't support dimming (GenLevelCtrl cluster)?

No does not work. How should it. The cluster is not even available on the ZBMINIR2 and you can't bind it to other decices,

**Comment by andrei-lazarov** [comment]
@uschindler 
It's very much a fault!
**It toggles itself even if it's not part of the bound group.**

**Comment by uschindler** [comment]
> [@uschindler](https://github.com/uschindler) It's very much a fault! **It toggles itself even if it's not part of the bound group.**

OK, I did not test this, because my wanted behaviour was to let it switch its own relay in addition to other lamps (see my previous comment).

I have no idea how that happens. Could be some bug, but that's not relevant for me. My main purpose was to synchronize other lights. The reason for this in your case it maybe some other binding in parallel. I had the same thing in ZHA: It was still bound to a group that was already deleted in the HASS UI, but the ID still existed on the devices. After cleaning that up, it worked well. Maybe it is something similar. The problem is that you can't easily see which bindings a device currently has (at least not in ZHA; z2m may be different), so I was watching the transmission logs - and noticed that the devive sent another group ID when toggling leftover from a previous test.

Maybe other option is to factory reset the device and reconfigure before trying again.

P.S.: You can also synchronize toggling of multiple ZBMINR2 when you put them all in same group and bind themselves to this group. Of course if they are async (one is on, other is off) then each button press toggles between them. 🤣 Nice effects, but absolutely how it should behave (although not always wanted).
